### ISSUE TITLE
second level dependencies exportOverrides

### ISSUE THREAD

**Issue by bpletzer** [yatskevich/grunt-bower-task#17]
Hi

grunt-bower-task is our dependency manager of choice and works quite well.

Our structure has a src-folder, bower dependencies get installed to src/vendor - i.e. tracking gets installed to src/vendor/tracking (built files are stored in dist/)

To archieve the folder structure we are using exportsOverride. For tracking it could be 

tracking-Package using anotherPackage:

``` javascript
"anotherPackage": {
            "": "dist/"
        }
```

so that only built files from the dist folder get copied to src/vendor/anotherPackage/

The problem arises when someone else is including the tracking-package - one would only use exportsOverride for that:

secondLevel-Package using tracking

``` javascript
"tracking": {
            "": "dist/"
        }
```

But now the information for anotherPackage is not set. The owner of the secondLevel-Package does not even know that tracking is depending on anotherPackage. Unfortunatelly now anotherPackage gets installed to src/vendor/anotherPackage, but all its contents, not just the dist-folder => referencing anotherPackage from tarcking does not work anymore, since tracking is assuming to find the main file in src/vendor/anotherPackage, not in src/vendor/anotherPackage/dist

Any ideas beside to state sub-dependencies explicitly?

Greetings
Benedikt

p.s.: We are using require-js on the client-side. First Level Dependencies always are referenced via vendor/anotherPackage which works in developement as well as when the package is a dependency itself:

tracking => vendor/anotherPackage
hook => vendor/tracking => vendor/anotherPackage

**Comment by bpletzer** [comment]
For our own packages I reallywould likeuse the main statement in component.json, unfortunatelly twittersays otherwise - main should always be only file for every extension. Doen't sound hat liberal for me

For third party dependencies the problem would stay the same, since jquery for example does not rpovide a component.json

Any guesses so far?

**Comment by yatskevich** [comment]
@bpletzer I don't quite get your problem. If you can provide some sample `component.json` with publicly available packages then I'll be able to reproduce your issue locally and think about it more deeply. BTW, version 0.2.0 of the task was released yesterday. Take a look at `layout` option.

**Comment by bpletzer** [comment]
I hope the following example illustrates my problem.

First of all you see the bower configuration ( **Gruntfile.js** ) for a self written package, which is the same as the app cosuming that package. That is, copy the main component files to src/components/[component]/

``` javascript
///...
    bower : {
      install : {
        options : {
          targetDir : 'src/components/',
          cleanup : true
        }
      }
    },
//...
```

Next, one sees **package/component.json** that includes itself a third party library, which has no component.json itself. So one has to use exportsOverides to point to the main file bin/lib.js

``` javascript
//...
 "name": "package",
 "main": ['src/main.js'],
 "dependencies": {
    "library": "",
  },
  "exportsOverride": {
    "library": {
      "": "bin/lib.js"
    },
//...
```

Now the problem arises. In **app/component.json** package gets named as dependency. But not the library, which nevertheless gets installed, since it's the second level dependeny, needed by package

``` javascript
//...
 "name": "app",
 "dependencies": {
    "package": ""
  },
//...
```

The final result is shown in the following tree structure. Package has a components-folder with **package/components/library/ _bin_ /lib.js** and a **package/src/components/library/lib.js**

App, in contrast, has a components-folder with **app/components/library/ _bin_ /lib.js** and a **app/src/library/ _bin_ /lib.js**. The last one is the problem. The Third-Party Library is not where package expects it to be.

```
package
├── component.json
├── components
│   └── library
│       └── bin
│           └── lib.js
├── Gruntfile.js
└── src
    ├── components
    │   └── library
    │       └── lib.js
    └── main.js

app
├── component.json
├── components
│   ├── library
│   │   └── bin
│   │       └── lib.js
│   └── package
│       └── see.package.tree
├── Gruntfile.js
└── src
    ├── app.js
    └── components
        ├── library
        │   └── bin
        │       └── lib.js
        └── package
            └── main.js

```

```
package
└── src
    └── components
        └── library
            └── lib.js

VS.

app
└── src
    └── components
        └── library
            └── bin
                └── lib.js
```

This problem can only be tackled right now if app/component.json adds an exportsOverride for library. But that makes no sense. __ _The owner of app cannot know that library is a second-level dependency of package and needs special handling, a.k.a. exportsOverrides_ __

So what could be done. Right know my only guess would be, that not only exportsOverrides of app/component.json gets parsed, but also for all its dependencies...

Hope that pointed out my problem

Thank you again! Despite that problem your task works well for me

Benedikt

**Comment by bpletzer** [comment]
still too complicated?

**Comment by yatskevich** [comment]
@bpletzer, you're right. The only way to do what you want is to explicitly declare app's overrides for every dependent library.

`grunt-bower-task` is open to contributions, so you can create a PR with implementation of the feature. 
We can discuss how to approach a solution here or via email.
